### PR TITLE
Added tests & improved documentation for beans_wrap_markup.

### PR DESCRIPTION
**Pull request by hellofromtonya**
1. Added unit tests.
2. Added integration tests.
3. Improved the function's DocBlock.

### CODE REVIEW

File: lib/api/html/functions.php [hunk 1]
```diff
@@ -389,18 +389,22 @@ function beans_reset_markup( $id ) {
...
  * @since 1.0.0
  *
- * @param string       $id         The markup ID.
+ * @param string       $id         The wrap markup's ID.
  * @param string       $new_id     A unique string used as a reference. The $id argument may contain sub-hook(s).
- * @param string       $tag        The HTML wrap tag.
+ * @param string       $tag        The wrap's HTML tag.
  * @param string|array $attributes Optional. Query string or array of attributes. The array key defines the
  *                                 attribute name and the array value define the attribute value. Setting
  *                                 the array value to '' will display the attribute value as empty
```
Comment (christophherr) [inline lib/api/html/functions.php]: Do we need tests for setting the value to `'', false, or null` or are those "standard" behaviours?

File: tests/phpunit/integration/api/html/beansWrapMarkup.php [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		);
+
+		// Clean up.
+		remove_action( 'foo_after_markup', array( $instance, 'callback' ), 1 );
+	}
+
+	/**
+	 * Test beans_wrap_markup() should not pass the given attributes to anonymous action.
+	 */
+	public function test_should_not_pass_attributes_for_after_markup_hook() {
```
[the remark this answers is not in the document]
Comment (hellofromtonya) [reply tests/phpunit/integration/api/html/beansWrapMarkup.php]: >Don't the previous and last test cover that no attributes are passed to the after markup hook?

Nope, the previous test focused on the registration process only, i.e. by checking if the anonymous action's callback did get registered to the `'foo_after_markup'` hook.

This test is checking specifically that the attributes are not being passed to the anonymous action.  How?  It compares what was stored in the anonymous action mock's property "callback" comparing it to what we expect to be stored:

```
$this->assertSame( $args, $anonymous_action_mock->callback );
```

>By anonymous action you mean $new_id was not passed
Nope, I mean the anonymous action.  Look at the code:

```
function beans_wrap_markup( $id, $new_id, $tag, $attributes = array() ) {
	$args = func_get_args();
	unset( $args[0] );

	_beans_add_anonymous_action( $id . '_before_markup', array( 'beans_open_markup', $args ), 9999 );

	unset( $args[3] );

	_beans_add_anonymous_action( $id . '_after_markup', array( 'beans_close_markup', $args ), 1 );

	return true;
}
```

Notice that it is calling the function `_beans_add_anonymous_action()`.  That function creates an instance of the anonymous action.  

In the test suite, I'm mocking (or stubbing) that functionality.

But the terminology within Beans is: "anonymous action".

File: lib/api/html/functions.php (hunk 1 above)
Comment (hellofromtonya) [reply lib/api/html/functions.php]: The tag by itself does not the current functionality of this function.  

That said, if there is no tag, then technically we should bail out.  Why? Nothing will get built, i.e. no HTML will be rendered for the wrap.   Thinking through that scenario, I don't think we run into much risk of adding a bail out.

I'll do that now.

File: tests/phpunit/integration/api/html/beansWrapMarkup.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (christophherr) [reply tests/phpunit/integration/api/html/beansWrapMarkup.php]: The unit test is great. It makes total sense.

To me, the integration test doesn't explain itself clearly enough.

- The only reference to anonymous action is in the return parameter DocBlock of `get_instance_from_wp()` about 70 lines above. 

- There are only two differences to the previous test:
 1. $new_id is an empty string
 2. No check if the callback was registered ($this->assertTrue( has_action( 'foo_after_markup' ) );

- The term `anonymous action` is nowhere to be found in the test. 
My brain tries to fill the void by finding the most obvious difference -> an empty string passed -> missing parameter == anonymous action.

File: tests/phpunit/unit/api/html/beansWrapMarkup.php [hunk 3]
```diff
@@ -33,6 +33,15 @@ protected function setUp() {
 		require_once __DIR__ . '/fixtures/class-anonymous-action-stub.php';
 	}
 
+	/**
+	 * Test beans_wrap_markup() should false when an empty tag is given.
```
Comment (christophherr) [inline tests/phpunit/unit/api/html/beansWrapMarkup.php]: `should return false`

Comment (hellofromtonya) [reply tests/phpunit/unit/api/html/beansWrapMarkup.php]: Doh.

File: tests/phpunit/integration/api/html/beansWrapMarkup.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (hellofromtonya) [reply tests/phpunit/integration/api/html/beansWrapMarkup.php]: I see the confusion now.  The test `test_should_register_beans_close_markup_to_given_id_after_markup_hook` had a check in it for the callback property.  That's wrong.

To improve readability, I moved the `wp_filter` tests into each integration method.

>The term anonymous action is nowhere to be found in the test.

Correct. But when you look at the function's code, it's used multiple times.  The function is calling another Beans' function that creates the anonymous function.  It is the `$instance` that we are checking.  Would you prefer `$instance` to be renamed to `$anonymous_action`?  Would that help?

[the remark this answers is not in the document]
Comment (hellofromtonya) [reply tests/phpunit/integration/api/html/beansWrapMarkup.php]: Okay, here's what I did:

1. I renamed the `$instance` to `$anonymous_action` to better represent what the object is.
2. The intent of `test_should_not_pass_attributes_for_after_markup_hook` test is to check that the `unset( $args[3] );` in the `beans_wrap_markup()` function works as intended.  Therefore, I simplified the test to _only_ check that the attributes do not exist in the anonomyous action, i.e.meaning they were not stored in the `$callback` property.
3. I moved the `$wp_filter` tests down into the appropriate methods so that you can see them instead of having to scroll up to them.

Let me know if that provides better clarity.

[the remark this answers is not in the document]
Comment (christophherr) [reply tests/phpunit/integration/api/html/beansWrapMarkup.php]: Thank you.
Sorry for not being clearer in my initial comments. 

1. If we want to point out that it is an instance, `$anonymous_action_instance` is longer but  makes sense to me. Otherwise, it's great. 
2. I think this was the main point of my `why is the test here?`  I like tests that will tell us clearly what we broke.
3. That's an added bonus :) If you'd rather keep it in the `get_instance_from_wp()` method to be more DRY, that's fine by me.

[the remark this answers is not in the document]
Comment (hellofromtonya) [reply tests/phpunit/integration/api/html/beansWrapMarkup.php]: Not a problem.  Your points were spot on and valid.  It should took me a couple of iterations to figure out how to achieve what you were looking for.

1. I think `$anonymous_action` is fine.  It's an object.  I don't think we need to go so far as to declare that it's one.  We are using it as an object within each test.  Therefore, it should be clear as to its data type.
2. Cool.  I didn't want us to loose the fact that we do the unset.  It can get lost in the details.  Should we break that down the road, it might not be clear why without that test.
3. True. But you're right.  It's more clear now.  Readability is important.

Thanks for your help on this one!  You did a great job!
